### DISCUSSION TITLE
How to send a RTU broadcast message

### DISCUSSION THREAD

**Discussion by robotiko** [eModbus/eModbus#323] [Q&A] [answered]
Hi, I'm trying to use RTU client to send a broadcast message. 
request has broadcast address protected to not use it but cannot get  the addBroadcastMessage to work with Message.

can  provide an example on how to use RTU  client broadcast message?
how to map Message to the params of  addBroadcastMessage(const uint8_t *data, uint8_t len) ?

thanks

**Comment by Miq1** [comment]
If you will have a [look at the documentation](https://emodbus.github.io/modbusclient-rtu-api#error-addbroadcastmessageconst-uint8_t-data-uint8_t-len) you will learn the ``addBroadcastMessage()`` does prefix your data with a ``0x00`` broadcast server ID and add the required CRC - nothing else. There is nothing to be done about the data, as it completely will depend on your respective requirements and is not standardized at all.

**Reply by robotiko** [reply]
Hi Miq1, 
thank you for your fast reply. Yes I checked the documentation and the code. 

The broadcast message reception at server (registerBroadcastWorker) works great handling the message contents so I want to keep the Message entity properly sent. (tested with ModBus Poll)
However the client interface for broadcast  is not using the Messsage entity. Basically the params is to add whatever content  you want (some size limits, etc ), but if I want to send a Message, have to decompose in its contents and seems that  I'm missing something here becuase dont get the broadcast message at Server. 

  // Define message
     ModbusMessage msg1;
     // Set it to a 0x03 read holding register request
     Error err = msg1.setMessage(1, WRITE_HOLD_REGISTER, address, value);
     if (err!=SUCCESS) {
      ModbusError e(err);
      LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
      }
      err = MB.addBroadcastMessage(msg1.data(),msg1.size());


So basically my question resumes as how can I use the  addBroadcastMessage to send a Message object?


Thanks a lot in advance

**Reply by Miq1** [reply, marked as answer]
This is in fact a design decision. A proper ``ModbusMessage`` always must have a valid server ID, but the broadcast ID of 0x00 is not a valid one according to the Modbus standard.

As a way around that you could construct a ``ModbusMessage msg`` as usual, but give the ``msg.data()+1`` pointer and ``msg.size()-1`` value to the broadcast call.

**Reply by robotiko** [reply]
Thanks a lot @Miq1
